Add a guide about adding a SGB border (#70)
gbdev/gbdev.github.io: website/guides/sgb_border.md

# Adding an SGB border

This document was written to aid developers of DMG compatibile homebrew with adding Super Game Boy borders

I cannot guarantee skipping certain parts won't leave you confused, so if you wish to add SGB borders to already SGB capable software (that lacks borders), try not to skip anything

> 

Reviewer: Good idea, other guides have a similar introduction (not necessarily in bullet points)

## Enabling SGB features

Before we can do anything, we must first specify in the header that this game is an SGB game.

> To enable SGB features: <!-- i think this is worded a little awkwardly, but i cant exactly think of how to fix it >< -->

Reviewer 2: I wanted to mention that what controls the "enabling of SGB features" is the ROM's header, and not e.g. code. I figured that it could help with troubleshooting?

> To enable SGB features: <!-- i think this is worded a little awkwardly, but i cant exactly think of how to fix it >< -->

Reviewer: Uh, okay, I see your point.. We should then try to simplify it a bit (e.g. to help restrict the cases in which really this "troubleshooting" may happen and the user should be aware of the fact it may involve SGB side, if I understood you correctly)

## Uploading

As stated before, the SGB border consists of tile data, picture data, and palette data. these are split across 2-3 packets:

- `CHR_TRN` (`$99`) is used to send 4KiB of tile data.

  - since the border can use up to 8KiB of tiles, bit 0 of the second byte specifies which "half" Youre sending
    - `$99, $00` if the screen is loaded with the first 4KiB of tile data
    - `$99, $01` if the screen is loaded with the second 4KiB of tile data

- `PCT_TRN` (`$a1`) is used to send the picture and palette data. it also swaps the border, generally a good idea to send it after the tile data[^1]
  - assuming tiles `0`-`255` use VRAM from `$8000` to `$8fff`:
    - the picture data must be at `$8000`-`$873f` (last 64 bytes are _usually_ offscreen, see [notes](#notes))
    - palette data must be at `$8800`-`$885f`
    - everything else is ignored
      - how You skip putting data at `$8740`-`$87ff` is up to You, I prefer doing separate copies, valen prefers copying tilemap and palette data in one go, with the area between them padded with `$ff`.

> If this makes no sense You could also read [pandocs](https://gbdev.io/pandocs/SGB_Command_Border.html)

Author: whoops forgot about this suggestion ><

[^1]: You can send a `CHR_TRN` up to \~60 frames after the `PCT_TRN` for it to apply to the current border, but not all emulators will emulate this. its fine to just pretend `CHR_TRN`s must go before `PCT_TRN`

## Notes

1. You can set the first row of tiles to Your transparent color to force superfamiconv to put the transparent tile as the 1st tile, however You must then exclude 64 bytes of the tilemap (`incbin "border.pct"` -> `incbin "border.pct", 64`)

2. You can use palettes 0-3 and 7, if You really know what Youre doing (animated borders yay), but You will probably have to edit the border in YY-CHR, as there arent really any other tools for that

3. when the SNES lags, scanline 225 of the SGB border will be visible! You can set the topmost row of the 29th row of tiles to black to hide this

> 4. if this doesnt work for You, or something here is unclear, then please contact me, preferably via discord (`@zlago`) or mastodon (`@zlago@mastodon.gamedev.place`) as I tend to check those reasonably often

Reviewer: we should just redirect to discord here (link to the chat channels on the gbdev.io website)

# Adding a custom SGB border

> This document was written to aid developers of DMG compatibile homebrew with adding Super Game Boy borders.
> 
> We wil see how:

Reviewer 3: ```suggestion
This document was written to aid developers of DMG compatible homebrew with adding Super Game Boy borders.

We will see how to:
```

- Detect whether we are running on a SGB
- Transfer the border's tiles
- Transfer and display the border's tilemap and palettes

<small>

Written by [sylvie (zlago)](https://zlago.github.io/me/), idea (and minor help) by [valentina (coffee bat)](https://coffeebat.neocities.org/), reviews and improvements by [ISSOtm](https://eldred.fr) and [avivace](https://github.com/avivace).

</small>

## Enabling SGB features

> Before we can do anything else, we must first specify in the header that this game is an SGB game.
> Otherwise, the SGB BIOS will ignore any packets we send, and we won't even be able to detect when we are running form a SGB.

Reviewer 3: ```suggestion
Before we can do anything else, we must first specify in the header that this game is aware of SGB features.
Otherwise, the SGB BIOS will ignore any packets we send, and we won't even be able to detect when the program is running on SGB.
```

To enable SGB features:

- The SGB flag (`$0146`) must be set to `$03`;
- The old licensee code (`$014b`) must be set to `$33`.

This can be achieved by passing the `--sgb-compatible` and `--old-licensee 0x33` flags to [`rgbfix`](https://rgbds.gbdev.io/docs/rgbgfx.1).

## Packets

The SGB BIOS can be "talked to" via *command packets*, sent bit by bit via the `P1`/`JOYP` register.

An SGB packet consists of:

- a "reset pulse" (`P1`=`%xx00xxxx`)
- 128 data pulses (`P1`=`%xx01xxxx` for "1", `P1`=`%xx10xxxx` for "0")
- a "0" pulse (`P1`=`%xx10xxxx`)

You must set `P1` to `%xx11xxxx` between each pulse.

This adds up to 16 Bytes of data (LSB first), If a packet doesn't read all 16 bytes, the unused bytes are ignored.

**You should wait 4 frames between each packet.**

For an example of such routine, see this [code](https://github.com/zlago/violence-gbc/blob/11cfdb6ee8a35e042fa9712484d814e0961cea7c/src/sub.sm83#L413-L463) and the related Pan Docs entry: [SGB Command Packet on Pandocs](https://gbdev.io/pandocs/SGB_Command_Packet.html).

> This guide glosses over a minor detail, as certain packets can be (albeit unccomon) more than 16 bytes.

Reviewer: Minor: should additional references/links be put here (e.g. a use case/example of sending more than 16 bytes?) or is it irrelevant/superficial?

## TRN

> Some packets will tell the SGB to copy the contents of the screen somewhere, including the packets needed for SGB borders. <!-- this could be reworded? -->

Reviewer: This needs to be reworded. Maybe mention what do we mean by "TRN" here?

> Some packets will tell the SGB to copy the contents of the screen somewhere, including the packets needed for SGB borders. <!-- this could be reworded? -->

Reviewer 3: ```suggestion
Bulk transfer (TRN) packets tell the SGB to copy the contents of the screen to buffers in Super NES work RAM. The `CHR_TRN` and `PCT_TRN` packets are used to send data for SGB borders.
```

For that to function properly, you must:

1. set `BGP` to `$e4` and `LCDC` to `$91` (screen enabled, BG uses tiles `$8000`-`$8fff` and tilemap `$9800`, WIN and OBJ disabled, BG enabled)
2. set `SCX` and `SCY` to `$00`

> 3. the tilemap consists of `$00`, `$01`..`$13`, 12 bytes padding (offscreen), `$14`..`$27`, padding, repeat untill `$ff` (inclusive)

Reviewer 3: ```suggestion
3. the tilemap consists of `$00`, `$01`..`$13`, 12 bytes padding (offscreen), `$14`..`$27`, padding, repeat until `$ff` (inclusive)
```

4. the data you want to send must be loaded at `$8000`-`$8fff`

> You can do 1, 2 and 3 via [this snippet](https://github.com/zlago/snek-gbc/blob/baef0369f57d2b0d58316cb1c28c6cc22475a6c9/code/init.sm83#L208-L230)

Reviewer: Should we include the snippet here as we did with the other one?

> - **You must load the data into VRAM before sending the TRN packet**
> - **You must wait ~8 frames after each TRN** (doesn't "stack" with the 4 frame packet delay)

Reviewer 3: ```suggestion
- **You must load the data into VRAM and enable the screen before sending the TRN packet**
- **You must wait ~8 frames after each TRN** instead of just 4
```

## Detecting SGB

Here's how a SGB detection routine should look like:

- wait 12 or more frames after the console boots
- send a `MLT_REQ`, selecting 2 or 4 players (`$89, $01` or `$89, $03`), and wait at least 1 frame

> - attempt to advance the read player (set `P1.5` to 0 then 1)

Reviewer 2: This sounds confusing. What does it mean?

Reviewer 3: My guess:
1. wait for the SGB to finish initializing
2. set multiplayer mode and wait for the SGB to recognize the request to go to multiplayer mode
3. test whether multiplayer mode was activated by reading the controller a couple times and seeing whether the input register reflects a player other than player 1
4. optionally turn off multiplayer mode

Reviewer: @ISSOtm

> - attempt to advance the read player (reset then set `P1.5`)
> - set `P1.4`-`P1.5` to `%11` (`%xx11xxxx`)
> - read `P1.0`-`P1.3`, and check if it either
> 	* has changed
> 	* isn't `%1111`
> - repeat a few times and branch somewhere if the test keeps failing

Reviewer 3: ```suggestion
- attempt to advance the read player (set `P1.5` to 0 then 1)
- set `P1.4`-`P1.5` to `%11` (`%xx11xxxx`)
- read `P1.0`-`P1.3`, and check if it either
	* has changed
	* isn't `%1111`
- advance and reread a few times and branch somewhere if the test keeps failing
```
"reset" is less clear than "set to 0", and "repeat a few times" could mean repeat the `MLT_REQ` too

ldh [rLYC], a
ld a, STATF_LYC
ldh [rSTAT], a
ld a, IEF_STAT
ldh [rIE], a
ei
	; wait for SGB
	ld b, 12 /* You must wait 12 or more frames before trying to send a packet */
	:halt
	dec b
	jr nz, :-
	; enable multi
	ld hl, Packets.mlt /* send a `MLT_REQ` */
	call Packet
	rept 4 /* then wait 4 more frames, since the SNES won't "read" the packet instantly */
		halt
	endr
	; check if SGB responds
		/* and now We actually try to detect the SGB
		setting `P1.5` to low then high advances the "read player"
		setting `P1.4` and `P1.5` high will make the SGB return which player is currently selected in `P1.0` and `P1.1` */
		lb bc, 5, LOW(rP1) /* b is loaded with 5 which is how many times We try to check if the player changes to anything but P1 */
		:ld a, P1F_4 /* try to advance player */
		ldh [c], a
		ld a, P1F_4|P1F_5 /* then set P1 to return the current player */
		ldh [c], a
		ldh a, [c]
		dec b
		jp z, .init ; give up /* if 5 (4 actually) attempts fail, assume this isn't an SGB */
		cp $ff /* `P1.6` and `P1.7` always return %1, We set `P1.4` and `P1.5` to %1, and DMG, or SGB player 1, return %1111 in `P1.0`-`P1.3` */
		jr z, :- ; /* try again if detection fails */

> ```

Reviewer 2: `sm83asm` is not recognised by any highlighter out there; I don't know which one is used on gbdev.io, but `rgbasm` is supported by highlight.js [via a plugin](https://github.com/gbdev/highlightjs-rgbasm), and alternatives like `avrasm`, `z80asm`, `x86asm`, or `nasm` degrade well enough.

Also, "block" comments (`/* */`) are very uncommon in assembly syntaxes; single-line comments (`;`) should be used instead, especially if using one of the alternatives listed above.

Reviewer: Agree on both points here

> If you adopt this, remember to change `.init` to wherever code should jump to if it's _not_ on SGB, and put code meant to run on SGB after the snippet.

Reviewer 2: The label should probably be called `.notSgb` then.

### SGB detection notes

- It would be a good idea to save somewhere in RAM whether the game is running on an SGB capable device or not, such that if you wish to change the border mid-gameplay, you won't have to perform SGB detection again

- **If you wish to only use 1 controller for the game, you will have to send another `MLT_REQ` to disable multiplayer** (`$89, $00`)

- You can change the waitloop to not use interrupts, or to use di+halt instead

## Border limitations

An SGB border has:

- 255 tiles + 1 transparent tile (preferably tile #0)
- 3 palettes of 15 (+ 1 transparent) colors, (up to 45 solid colors total)
- a 256x224px tilemap (there's a bit more to this, see [notes](#notes))

## Converting borders

With a recent version of [superfamiconv](https://github.com/Optiroc/SuperFamiconv):

```sh
superfamiconv -v -i input.png -p output.pal -t output.4bpp -P 4 -m output.pct -M snes --color-zero 0000ff -B 4
```

> - `--color-zero` should be the color _you_ used for transparency, in my case it was blue.
> 	* it can also be set to `00000000` to use the actual transparent color, however, this may cause some issues..
> - `-v` is optional

Reviewer 3: ```suggestion
- `--color-zero` should be the color that your image for transparency, in my case it was blue.
	* If your image has an alpha channel, it can be set can also be set to `00000000` to use the actual transparent color; however, this may cause some issues.
- `-v` is optional, for showing details of the conversion process
```

## Uploading borders

As stated before, the SGB border consists of tile data, picture data, and palette data. These are split across 2-3 packets:

- `CHR_TRN` (`$99`) is used to send 4KiB of tile data.

	* since the border can use up to 8KiB of tiles, bit 0 of the second byte specifies which "half" you're sending
		- `$99, $00` if the screen is loaded with the first 4KiB of tile data
		- `$99, $01` if the screen is loaded with the second 4KiB of tile data

- `PCT_TRN` (`$a1`) is used to send the picture and palette data. it also swaps the border, generally a good idea to send it after the tile data[^1]

	* assuming tiles `0`-`255` use VRAM from `$8000` to `$8fff`:
		- the picture data must be at `$8000`-`$873f` (last 64 bytes are _usually_ offscreen, see [notes](#notes))
		- palette data must be at `$8800`-`$885f`
		- everything else is ignored
			* how you skip putting data at `$8740`-`$87ff` is up to you, I prefer doing separate copies, valen prefers copying tilemap and palette data in one go, with the area between them pa

See also the related Pan Docs entry: [SGB Command Border](https://gbdev.io/pandocs/SGB_Command_Border.html).

> [^1]: You can send a `CHR_TRN` up to \~60 frames after the `PCT_TRN` for it to apply to the current border, but not all emulators will emulate this. its fine to just pretend `CHR_TRN`s must go before `PCT_TRN`

Reviewer 2: ```suggestion
[^1]: You can send a `CHR_TRN` [up to \~60 frames](https://github.com/pinobatch/little-things-gb/blob/master/sgbears/docs/long_story.txt) after the `PCT_TRN` for it to apply to the current border, but not all emulators will emulate this. It's fine to just pretend `CHR_TRN`s must go before `PCT_TRN`.
```

Reviewer 3: I might change the name of `long_story.txt` in a future revision of this repository. Sometimes I leave things as `.txt` while writing them until I finish enough of the first draft to be able to tell whether or not I'll be using features that make the Markdown source code less readable in a text editor. Should this tutorial link to `long_story.txt` in a particular revision, or should it wait until I've made a decision about renaming it?

Reviewer: The link right now it's broken for me. We should link to it in a particular git commit so the link won't break and you can PR the new link if/when it changes.

Reviewer 3: Link to a specific commit <https://github.com/pinobatch/little-things-gb/blob/b11b554d73c48a0f54fee0df31e59eb83806fcb4/sgbears/docs/long_story.md>

Reviewer 2: Thanks Pino! Updated my original suggestion accordingly.

## Notes

1. You can set the first row of tiles to your transparent color to force superfamiconv to put the transparent tile as the 1st tile, however you must then exclude 64 bytes of the tilemap (`incbin "border.pct"` -> `incbin "border.pct", 64`)

> 2. You can use palettes 0-3 and 7, if You really know what Youre doing (animated borders yay), but You will probably have to edit the border in YY-CHR, as there aren't really any other tools for that

Reviewer 3: ```suggestion
2. You can use palettes 0-3 and 7, if you really know what you're doing (animated borders yay). You will probably have to edit the border in YY-CHR, as there aren't really any other tools for that.
```
Unlike some other languages, English doesn't capitalize "you" as if it were a proper noun.

> 2. You can use palettes 0-3 and 7, if you really know what you're doing (animated borders yay). You will probably have to edit the border in YY-CHR, as there aren't really any other tools for that.

Reviewer 2: What does "if you really know what you're doing" imply?

Reviewer 3: There may be advanced techniques to draw parts of the border with the game palette (0) instead of the palettes reserved for the border (4-6). Currently no converter supports doing this. Would it be worthwhile to make a test ROM to confirm that it's possible?

Author: since its not that practical to use the game palette, it should be a low priority

Reviewer 2: Then I think this shouldn't be mentioned in the tutorial, beyond something like "palette 0 is used by the displayed GB screen, palettes 1-3 and 7 are used by the SGB BIOS and will be overwritten when opening various menus".

Mentioning that they *can* be used, but not what the limitations are, is likely to be more confusing.

> 3. when the SNES lags, scanline 225 of the SGB border will be visible! You can set the topmost row of the 29th row of tiles to black to hide this

Reviewer 2: ```suggestion
3. When the SNES lags, scanline 225 of the SGB border will be visible! You can set the topmost row of the 29th row of tiles to black to hide this.
```

> 4. if this doesn't work for you, ask for help on the [gbdev](https://gbdev.io/chat.html) channels

Reviewer 2: ```suggestion
4. If this doesn't work for you, you can ask for help on the [gbdev](https://gbdev.io/chat.html) channels.
```

# Adding an SGB border

This document was written to aid developers of DMG compatibile homebrew with adding Super Game Boy borders

I cannot guarantee skipping certain parts won't leave you confused, so if you wish to add SGB borders to already SGB capable software (that lacks borders), try not to skip anything

## Enabling SGB features

Before we can do anything, we must first specify in the header that this game is an SGB game.

To enable SGB features: <!-- i think this is worded a little awkwardly, but i cant exactly think of how to fix it >< -->

- SGB flag (`$0146`) must be set to `$03`
- Old licensee code (`$014b`) must be set to `$33`

## Packets

> Communicating with the SGB is done by bitbanging "packets" through the `P1` register

Reviewer 3: Is it that [bit banging](https://en.wikipedia.org/wiki/Bit_banging) is not well known as a term of art for toggling GPIOs, or some other reason?

> Communicating with the SGB is done by bitbanging "packets" through the `P1` register

Author: ..."Communicating with the SGB is done by sending a bitstream through the `P1` register"?

> Communicating with the SGB is done by bitbanging "packets" through the `P1` register

Reviewer: > Is it that [bit banging](https://en.wikipedia.org/wiki/Bit_banging) is not well known as a term of art for toggling GPIOs, or some other reason?

"Bit banging" is jargon, which should be avoided when not necessary

> Communicating with the SGB is done by bitbanging "packets" through the `P1` register

Author: does LSB/MSB not count as jargon? would you approve of `sent bit-by-bit ("bitbanged")` or `[bitbanged](wikipedia link)`?

You must set `P1` to `%xx11xxxx` between each pulse.

> This adds up to 16 Bytes of data (LSB first), If a packet doesn't read all 16 bytes, the unused bytes are ignored.

Reviewer 3: It has a comma splice after the change.
```diff
-first), if
+first). If
```

> This adds up to 16 Bytes of data (LSB first), If a packet doesn't read all 16 bytes, the unused bytes are ignored.

Author: this is the type of thing that makes me go
```
git ci --amend --no-e
git fpush
```
then
```
git add -u
git ci --amend --no-e
git fpush
```
after i realise i forgot to stage ><
